src/update_engine/payload_processor.cc

```diff
@@ -0,0 +1,445 @@
+      return _retval;                                           \
+    }                                                           \
+  } while (0);
+
+ActionExitCode PayloadProcessor::VerifyPayload(
+    const std::string& update_check_response_hash,
+    const uint64_t update_check_response_size) {
+  LOG(INFO) << "Verifying delta payload using public key: " << public_key_path_;
+
+  // Verifies the download size.
+  TEST_AND_RETURN_VAL(kActionCodePayloadSizeMismatchError,
+                      update_check_response_size ==
+                      manifest_metadata_size_ + buffer_offset_);
+
+  // Verifies the payload hash.
+  const string& payload_hash_data = hash_calculator_.hash();
+  TEST_AND_RETURN_VAL(kActionCodeDownloadPayloadVerificationError,
+                      !payload_hash_data.empty());
+  TEST_AND_RETURN_VAL(kActionCodePayloadHashMismatchError,
+                      payload_hash_data == update_check_response_hash);
+
+  // Verifies the signed payload hash.
+  if (!utils::FileExists(public_key_path_.c_str())) {
+    LOG(WARNING) << "Not verifying signed delta payload -- missing public key.";
+    return kActionCodeSuccess;
+  }
+  TEST_AND_RETURN_VAL(kActionCodeSignedDeltaPayloadExpectedError,
+                      !signatures_message_data_.empty());
+  vector<char> signed_hash_data;
+  TEST_AND_RETURN_VAL(kActionCodeDownloadPayloadPubKeyVerificationError,
+                      PayloadSigner::VerifySignature(
+                          signatures_message_data_,
+                          public_key_path_,
+                          &signed_hash_data));
+  OmahaHashCalculator signed_hasher;
+  TEST_AND_RETURN_VAL(kActionCodeDownloadPayloadPubKeyVerificationError,
+                      signed_hasher.SetContext(signed_hash_context_));
+  TEST_AND_RETURN_VAL(kActionCodeDownloadPayloadPubKeyVerificationError,
+                      signed_hasher.Finalize());
+  vector<char> hash_data = signed_hasher.raw_hash();
+  PayloadSigner::PadRSA2048SHA256Hash(&hash_data);
+  TEST_AND_RETURN_VAL(kActionCodeDownloadPayloadPubKeyVerificationError,
+                      !hash_data.empty());
+  if (hash_data != signed_hash_data) {
+    LOG(ERROR) << "Public key verification failed, thus update failed. "
+        "Attached Signature:";
+    utils::HexDumpVector(signed_hash_data);
+    LOG(ERROR) << "Computed Signature:";
+    utils::HexDumpVector(hash_data);
+    return kActionCodeDownloadPayloadPubKeyVerificationError;
+  }
+
+  return kActionCodeSuccess;
+}
+
+bool PayloadProcessor::GetNewPartitionInfo(uint64_t* rootfs_size,
+                                         vector<char>* rootfs_hash) {
+  TEST_AND_RETURN_FALSE(manifest_valid_ &&
+			manifest_.has_new_partition_info());
+  *rootfs_size = manifest_.new_partition_info().size();
+  vector<char> new_rootfs_hash(manifest_.new_partition_info().hash().begin(),
+                               manifest_.new_partition_info().hash().end());
+  rootfs_hash->swap(new_rootfs_hash);
```

Comment:
This looks like some rootfs occurrences slipped through the s/rootfs/partition/g done in 6664398?